hadoop-hdds/docs/content/feature/Snapshot.md

```diff
@@ -0,0 +1,68 @@
+ * backup/restore: Create hourly/daily/weekly/monthly snapshots for backup and restore.
+ * Archival and Compliance: Take snapshots for compliance purpose and archieve them as required.
+ * Replication and DR : Snapshots provide a frozen immutable image of the bucket on the source Ozone cluster. Snapshots can be used for replicating these immutable bucket images to remote DR site.
+ * Incremental Replication : Snapshots feature with efficent snapshot-diff can offer an effcient way to incremntally sync up remote copy of the bucket.
+
+## Snapshot APIs
+
+Snapshot feature is available through 'ozone fs' and 'ozone cli'. This feature can also be programattically accessed from native java Ozone object store client. The feature provides following functionalities :
+* Create Snapshot: Create an instantenous snapshot for a given bucket
+```shell
+ozone sh snapshot create [-hV] <bucket> [<snapshotName>]
+```
+* List Snapshots: List all snapshots of a given bucket
+```shell
+ozone sh snapshot list [-hV] <bucket>
+```
+* Delete snapshot: Delete a given snapshot for a given bucket
+```shell
+ozone sh snapshot delete [-hV] <bucket> <snapshotName>
+```
+* Snapshot Diff: Given two snapshots, list all the keys that are different betwenn the them.
+```shell
+ozone sh snapshot diff [-chV] [-p=<pageSize>] [-t=<continuation-token>] <bucket> <fromSnapshot> <toSnapshot>
+```
+
+Snapshot diff is an asynchronous API. The first time the API is invoked, it starts a background process to calculate the Snapshot diff, and returns "Retry" with suggested duration for the retry operation. Once the snapshot diff is computed, this API returns the diffs in multiple Pages. With each Diff response, it also returns a continuation token to continue from the last page of diff results.  This API is safe to be called multiple times for a given Snapshot source and destination pair. Internally Ozone computes Snapdiff only once and stores it for future invocations of the same Snapshot Diff API. 
+
+## Architecture
+
+Ozone Snapshot architectue leverages the fact that data blocks once written, remain immutable for their life time. These data blocks are reclaimed only when the object-key/file metadata that references them, is deleted from the Ozone namespace. All of this Ozone metadata is stored on the OM nodes in the Ozone cluster. When a user takes a snapshot of an Ozone bucket, internally the system takes snapshot of the Ozone metadata in OM nodes. Since Ozone doesn't allow updates to datanode blocks, integrity of data blocks referenced by Ozone metadata snapshot in OM nodes remains intact. Ozone key deletion service is also aware of Ozone snapshots.  Key deletion service does not reclaim any key as long as it is referenced by the active object store bucket or any of its snapshot. When the snapshots are deleted, a background garbage collection service reclaims any key that will not be part of any snapshot or active object store.
```

Comment:
```suggestion
Ozone Snapshot architecture leverages the fact that data blocks once written, remain immutable for their lifetime. These data blocks are reclaimed only when the object key metadata that references them, is deleted from the Ozone namespace. All of this Ozone metadata is stored on the OM nodes in the Ozone cluster. When a user takes a snapshot of an Ozone bucket, internally the system takes snapshot of the Ozone metadata in OM nodes. Since Ozone doesn't allow updates to datanode blocks, integrity of data blocks referenced by Ozone metadata snapshot in OM nodes remains intact. Ozone key deletion service is also aware of Ozone snapshots.  Key deletion service does not reclaim any key as long as it is referenced by the active object store bucket or any of its snapshot. When the snapshots are deleted, a background garbage collection service reclaims any key that will not be part of any snapshot or active object store.
```